[inductor] Rewrite Triton templates + epilogue fusion

jansel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90739
* __->__ #90738

Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

jansel committed:
[inductor] Rewrite Triton templates + epilogue fusion



Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented:
@jansel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41474473).

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented:
@jansel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41474473).

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented:
@jansel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41474473).

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

voznesenskym commented:
+1500 / -1900 is really hard to review - is there any sane way to break this up? For example, `torch/_inductor/kernel/mm_common.py` feels like it could be its own PR, as could the TritonTemplate mm_template, bmm_template could be its own PR as well probably, maybe some utils could get refactored before this... 

Also equal number of test deletions and test additions feels like a net neutral in terms of test coverage, aside from testing algorithm selection, is there a good way to test the new templates / kernel parts?

jansel commented:
@jansel has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41474473).

jansel commented:
> +1500 / -1900 is really hard to review - is there any sane way to break this up? For example, `torch/_inductor/kernel/mm_common.py` feels like it could be its own PR, as could the TritonTemplate mm_template, bmm_template could be its own PR as well probably, maybe some utils could get refactored before this...
> 
> Also equal number of test deletions and test additions feels like a net neutral in terms of test coverage, aside from testing algorithm selection, is there a good way to test the new templates / kernel parts?

We want this change to be atomic since it is switching from the old template system to the new one.  I only added two kernels, so the change could be tested, more kernels in next diff in stack.

eellison commented on torch/_dynamo/testing.py:
```diff
@@ -231,8 +231,12 @@ def _fn(*args, **kwargs):
 
-def rand_strided(size, stride, dtype=torch.float32, device="cpu"):
-    needed_size = sum((shape - 1) * stride for shape, stride in zip(size, stride)) + 1
+def rand_strided(size, stride, dtype=torch.float32, device="cpu", extra_size=0):
```
what necessitated `extra_size` here ?

jansel replied:
It is needed for the case of `offset>0`.  We need the returned pointer to not include an offset, since the offset is already included in the generated code.

eellison commented on torch/_inductor/scheduler.py:
```diff
@@ -894,6 +875,12 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
         if node2.is_template():
             return False  # only epilogues
+        if node1.is_template() and (
+            node2.has_aliasing_or_mutation()
```
why do we want to skip if there is aliasing ?

jansel replied:
We may want to support this eventually, though for now being conservative of when to apply this.

eellison commented on torch/_inductor/utils.py:
```diff
@@ -30,6 +30,17 @@
+        raise NotImplementedError("requires Triton")
+
+
+def cdiv(x, y):
```
`ceildiv` already exists

jansel replied:
Thanks!  Will switch to it.

eellison reviewed (commented):
cool ! I didn't look too closely but LGTM

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

voznesenskym commented:
> > +1500 / -1900 is really hard to review - is there any sane way to break this up? For example, `torch/_inductor/kernel/mm_common.py` feels like it could be its own PR, as could the TritonTemplate mm_template, bmm_template could be its own PR as well probably, maybe some utils could get refactored before this...
> > Also equal number of test deletions and test additions feels like a net neutral in terms of test coverage, aside from testing algorithm selection, is there a good way to test the new templates / kernel parts?
> 
> We want this change to be atomic since it is switching from the old template system to the new one. I only added two kernels, so the change could be tested, more kernels in next diff in stack.

You can introduce new structures and have them live side by side, but then do the swap atomically in another PR. I don't care enough to argue the point tho.

voznesenskym commented on test/inductor/test_select_algorithm.py:
```diff
@@ -0,0 +1,146 @@
+            return torch.mm(a, b)
+
+        foo(
+            torch.randn(8, 32, device="cuda"),
+            torch.randn(32, 8, device="cuda"),
```
Do you mind adding a small comment about why this combination of inputs picks an algorithm, but the float64 skips?

jansel replied:
There already is a comment here:
https://github.com/pytorch/pytorch/blob/786c697e4727848a4ab0597c18330fad786adf5e/test/inductor/test_select_algorithm.py#L99

voznesenskym commented on torch/_inductor/codecache.py:
```diff
@@ -57,6 +58,36 @@ def cache_dir():
+    @staticmethod
+    @functools.lru_cache(None)
+    def _subdir():
+        subdir = os.path.join(cache_dir(), "cached_tunings")
+        os.makedirs(subdir, exist_ok=True)
+        return subdir
```
Did this throw on internal deployments? Or just work as is?

jansel replied:
It worked fine on a devserver.  This writes to `/tmp`

voznesenskym commented on torch/_inductor/codecache.py:
```diff
@@ -57,6 +58,36 @@ def cache_dir():
+        super().__init__()
+        self.unique_name = unique_name
+
+    def lookup(self, key, generate):
```
put types on this fn please

jansel replied:
will do

voznesenskym commented on torch/_inductor/codecache.py:
```diff
@@ -57,6 +58,36 @@ def cache_dir():
+        """
+        path = os.path.join(self._subdir(), code_hash(self.unique_name + repr(key)))
+        if not os.path.exists(path):
+            value = generate()
```
What type is value?

jansel replied:
It can be anything supported by `json`.  This is a generic caching class.

voznesenskym commented on torch/_inductor/codecache.py:
```diff
@@ -57,6 +58,36 @@ def cache_dir():
+        path = os.path.join(self._subdir(), code_hash(self.unique_name + repr(key)))
+        if not os.path.exists(path):
+            value = generate()
+            write_atomic(path, json.dumps(value))
+        return self._read_file(path)
```
You don't need to read a file and parse JSON if you just wrote it, consider just returning early in the `not os.path_exists` case

jansel replied:
While this is true, it could create some pretty subtle bugs since the returned value might be slightly different on a cache hit versus a cache miss.  Specifically json will change aliasing relationships.

voznesenskym commented on torch/_inductor/codecache.py:
```diff
@@ -85,14 +116,18 @@ def write(source_code, ext, extra=""):
     return basename, path
 
 
+def write_atomic(path, source_code):
```
types please

jansel replied:
ok

voznesenskym commented on torch/_inductor/codegen/common.py:
```diff
@@ -200,13 +199,29 @@ def _lookup(prefix, odict, name):
 
     def input(self, name):
-        name = V.graph.scheduler.mutation_real_name.get(name, name)
+        if V.graph.scheduler:
```
j/w when would we not have a scheduler here? Did something change about the lifecycles here?

jansel replied:
We run algorithm selection before scheduling.

voznesenskym commented on torch/_inductor/codegen/common.py:
```diff
@@ -547,8 +566,9 @@ def store(name, index, value, mode=None):
                     self.cse.store_cache[name] = value
-                    for other_name in self.current_node.get_mutations():
-                        self.cse.store_cache[other_name] = value
+                    if self.current_node:
```
same here, and below - how did these calls work before that we now have to change to None checking?

jansel replied:
We run algorithm selection before scheduling, so need to codegen benchmarking code before these are populated.

voznesenskym commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1176,6 +1197,9 @@ def can_fuse(self, node1, node2):
             if not (numel1 == numel2 and rnumel1 == rnumel2):
                 return False
 
+            if node1.is_template():
+                return True
```
Can you add a comment why a template means an early return?

jansel replied:
Sure

voznesenskym commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -312,6 +300,21 @@ def __init__(self):
         )
 
+        @functools.lru_cache(None)
+        def add_import_once(line):
+            self.header.writeline(line)
+
+        self.add_import_once = add_import_once
```
Where is add_import_once invoked?

jansel replied:
https://github.com/pytorch/pytorch/blob/786c697e4727848a4ab0597c18330fad786adf5e/torch/_inductor/select_algorithm.py#L286

voznesenskym commented on torch/_inductor/ir.py:
```diff
@@ -2666,12 +2734,25 @@ def codegen(self, wrapper):
         )
         self.output_view = output_view
         self.name = V.graph.register_buffer(self)
+        if kernel is not None:
+            self.kernel = kernel
+        if cpp_kernel is not None:
+            self.cpp_kernel = cpp_kernel
```
Is it ever valid to be in here with no kernels at all?

jansel replied:
Yes, the subclasses do that and define kernel as a class variable.

voznesenskym commented on torch/_inductor/ir.py:
```diff
@@ -2242,6 +2263,53 @@ def constant_to_device(self, device):
         return self.data.constant_to_device(device)
 
 
+class TemplateBuffer(Buffer):
```
Document this class

voznesenskym commented on torch/_inductor/kernel/mm_common.py:
```diff
@@ -0,0 +1,143 @@
+    return True
+
+
+def use_triton_template(layout):
+    return (
+        inductor_config.max_autotune
+        and layout.device.type == "cuda"
+        and layout.dtype in (torch.float16, torch.bfloat16, torch.float32)
+        and is_big_gpu(layout.device.index or 0)
+    )
```
is this specific to mm? If not, put it in a utils class?

jansel replied:
will do

voznesenskym commented on torch/_inductor/kernel/mm_common.py:
```diff
@@ -0,0 +1,143 @@
+    return f"tl.{dtype}".replace("torch.", "")
+
+
+@functools.lru_cache(None)
+def is_big_gpu(index):
+    cores = torch.cuda.get_device_properties(index).multi_processor_count
+    if cores < 80:  # V100
+        log.warning("not enough cuda cores to use max_autotune mode")
+        return False
+    return True
```
Why 80? - also consider using `torch.cuda.get_device_name` in the log.

jansel replied:
On smaller GPUs this gives "CUDA: Invalid argument" because the configurations use more register space that the GPU has.

80 cores is a V100, so this requires at least a V100.

voznesenskym commented on torch/_inductor/kernel/mm_common.py:
```diff
@@ -0,0 +1,143 @@
+        sympy.gcd(sym_k, config.kwargs["BLOCK_K"])
+        == config.kwargs["BLOCK_K"]
+    )
+    return dict(
+        GROUP_M=8,
+        EVEN_K=even_k_symbolic,
+        ALLOW_TF32=torch.backends.cuda.matmul.allow_tf32,
+        ACC_TYPE=acc_type(layout.dtype),
+        num_stages=config.num_stages,
+        num_warps=config.num_warps,
+        **config.kwargs,
+    )
```
Make this a real object? with a name and types?

bertmaher commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -312,6 +300,21 @@ def __init__(self):
+        self.add_import_once = add_import_once
+        self._metas = {}
+
+    def add_meta_once(self, meta):
```
I don't see `add_meta_once` used anywhere?

jansel replied:
https://github.com/pytorch/pytorch/blob/786c697e4727848a4ab0597c18330fad786adf5e/torch/_inductor/select_algorithm.py#L287

bertmaher commented on torch/_inductor/config.py:
```diff
@@ -156,11 +162,10 @@ class triton:
 
-    # use triton.autotune?
+    # use triton.autotune for pointwise ops with complex layouts
+    # this should only be disabled for debugging/testing
```
We should probably change the name of this to `autotune_pointwise` or something.  Maybe in a separate diff.

jansel replied:
will do

bertmaher reviewed (commented):
I have a few superficial comments but I haven't even got to the meat of the diff.  Tbh I don't think I'm going to be able to review this in a timely fashion, so I'll leave it to someone else to accept.

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

bertmaher reviewed (approved):
Gonna go ahead and approve.  I'm skimming, but realistically I don't want to block this since the perf wins are so big.

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

voznesenskym commented:
All my nits and concerns are addressed - I do not have the expertise to review the core kernel logic. Consider a post-land review w/ @ngimel  or maybe even @wickedfoo

voznesenskym commented:
You've murdered the tests, Jason

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

ngimel commented:
Looks like some test tolerances need to be adjusted

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented:
Perhaps rebasing onto https://github.com/pytorch/pytorch/commit/e2377c8300c578d648aac3bff6f96cdfce4308de will fix it.

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel committed:
Update on "[inductor] Rewrite Triton templates + epilogue fusion"


Differential Revision: [D41474473](https://our.internmc.facebook.com/intern/diff/D41474473/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D41474473/)!

cc mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jansel commented:
@pytorchbot merge
